Merge for "BIG task"

nikodrozd opened the pull request.

nikodrozd committed:
Added solution for "BIG task" (Flink, Spark and GeoJson HeatMap)

nikodrozd committed:
Added tests for "BIG task" (Flink, Spark and GeoJson HeatMap)

volodvro commented on src/main/scala/HeatMapMain.scala:
```diff
@@ -0,0 +1,28 @@
+
+object HeatMapMain extends App {
+  val outputFilePath = "src/main/resources/tweets.geojson"
+  val qtMaxLevel = 3
```
would be better to move configuration values to a separate file and load it at app start

nikodrozd replied:
Moved to config file

volodvro commented on src/main/scala/credentials/MongoDBCredentials.scala:
```diff
@@ -0,0 +1,14 @@
+
+object MongoDBCredentials {
+
+  val host = "localhost:27017"
```
move to config file please

nikodrozd replied:
Moved to config file

volodvro commented on src/main/scala/credentials/TwitterCredentials.scala:
```diff
@@ -0,0 +1,23 @@
+
+object TwitterCredentials {
+
+  private val consumerKey = "bBDkUPIwURjnkY39mK2sfUV72"
```
move to config file please

nikodrozd replied:
Moved to config file

volodvro commented on src/main/scala/utils/HeatMapGenerator.scala:
```diff
@@ -0,0 +1,59 @@
+    generateResultGeoJson(totalGeoJsonArray)
+  }
+
+  private[utils] def generatePoiTweetsGeoJson(tweets: Seq[Tweet]): JsArray = Spark.sc.parallelize(tweets, Spark.parallelizeLevel)
```
you should use mongo-spark-connector to load data from mongo

nikodrozd replied:
Changed to using mongo-spark-connector

volodvro commented on src/main/scala/utils/Spark.scala:
```diff
@@ -0,0 +1,11 @@
+import org.apache.spark.{SparkConf, SparkContext}
+
+object Spark {
+  val appName: String = "HeatMap Generator"
```
move to config file please

nikodrozd replied:
Moved to config file

volodvro commented on src/main/scala/model/QuadTreeGeo.scala:
```diff
@@ -0,0 +1,83 @@
+package model
+
+case class QuadTreeGeo(level: Int = 0,
```
Since Quad **always** has subquad, Option seems inapropriate for child representation. something with lazy initialization would be more descen

nikodrozd replied:
Reworked to use lazy initialization for child nodes

volodvro commented on src/main/scala/TweetCollectorMain.scala:
```diff
@@ -0,0 +1,22 @@
+  val streamSource: DataStreamSource[String] = env.addSource(new TwitterSource(TwitterCredentials.getTwitterCredentials))
+
+  val tweets = streamSource.filter(TweetStringFilter()).flatMap(StringToTweetFlatMap()).keyBy((tweet: Tweet) => tweet.id)
+  AsyncDataStream.unorderedWait(tweets, AsyncDatabaseResult(), 3000, TimeUnit.MILLISECONDS, 100)
```
100 was a proposed number to get familiar with the data format. This number is  supposed to be configurable, as well as  maximum quad tree level

nikodrozd replied:
Moved to config file

nikodrozd committed:
Code was refactored after review:
1) QuadTreeGeo child nodes changed from Option to lazy init
2) Data loading from MongoDB changed to mongo-spark-connector
3) Added config file
4) Test modified according to changes in code

nikodrozd committed:
Tweet collector parameters were moved to config file
